orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,519 @@
+    for node in fx_module.graph.nodes:
+        new_kwargs = {}
+        for k, v in node.kwargs.items():
+            if isinstance(v, torch.device):
+                v = v.type
+            new_kwargs[k] = v
+        node.kwargs = new_kwargs
+    for node in fx_module.graph.nodes:
+        if isinstance(node.target, torch._ops.OpOverload):
+            node.target = node.target.overloadpacket
+    fx_module.graph.lint()
+    fx_module.recompile()
+    return torch.jit.script(fx_module)  # type: ignore
+
+
+def _decorate_script_module(script_module: torch.jit.ScriptModule, expected_inputs, expected_outputs):
+    for i, input_value in enumerate(script_module.graph.inputs()):  # type: ignore
+        if input_value.debugName() == "self":
+            script_module.graph.eraseInput(i)  # type: ignore
+            break
+    for i, input_value in enumerate(script_module.graph.inputs()):  # type: ignore
+        input_value.setType(torch._C.TensorType.create_from_tensor(expected_inputs[i]))
+    for i, output_value in enumerate(script_module.graph.outputs()):  # type: ignore
+        output_value.setType(torch._C.TensorType.create_from_tensor(expected_outputs[i]))
+
+
+def _create_onnx_proto(script_module):
+    onnx_proto = _jit_graph_to_onnx_model(script_module.graph, torch.onnx.OperatorExportTypes.ONNX)
+    return onnx_proto
+
+
+def _create_onnx_model(onnx_proto):
+    return onnx.ModelProto.FromString(onnx_proto)
+
+
+def _create_onnx_session(onnx_proto, device):
+    # TODO(wechi): Add more EPs per PyTorch device types.
+    # TODO(wechi): enable external allocators.
+    if device.type == "cuda":
+        return onnxruntime.InferenceSession(onnx_proto, providers=["CUDAExecutionProvider"])
+    return onnxruntime.InferenceSession(onnx_proto, providers=["CPUExecutionProvider"])
+
+
+def _get_onnx_devices(values: Tuple[torch.Tensor]) -> Tuple[ORTC.OrtDevice]:  # type: ignore
+    assert all(value.device == values[0].device for value in values), "All values must be on the same device."
+
+    def _device_id_or_zero(device_id: int) -> int:
+        return device_id or 0
+
+    devices: Tuple[ORTC.OrtDevice] = tuple(  # type: ignore
+        ORTC.OrtDevice(  # type: ignore
+            _get_ort_device_type(value.device.type, _device_id_or_zero(value.device.index)),
+            ORTC.OrtDevice.default_memory(),  # type: ignore
+            _device_id_or_zero(value.device.index),
+        )
+        for value in values
+    )
+    return devices
+
+
+def _run_onnx_session_with_ortvaluevector(
+    sess: onnxruntime.InferenceSession,
+    input_names: Tuple[str],
+    inputs: Tuple[torch.Tensor],
+    input_devices: Tuple[ORTC.OrtDevice],  # type: ignore
+    output_names: Tuple[str],
+    outputs: Tuple[torch.Tensor],
+    output_devices: Tuple[ORTC.OrtDevice],  # type: ignore
+):
+    _nvtx_range_push("contiguous")
+    inputs = tuple(a.contiguous() for a in inputs)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("push_back_batch")
```

Comment:
It'd be helpful to explain why these few `_nvtx` lines are for the unfamiliar (like myself 😅 )

Reply:
Sure. You get
```
def _nvtx_range_push(name: str):
    """If PyTorch is installed with CUDA support, this starts NVTX range.

    Check torch.cuda.nvtx.range_push's document for more details.
    """
    if torch.cuda.is_available():
        torch.cuda.nvtx.range_push(name)


def _nvtx_range_pop():
    """If PyTorch is installed with CUDA support, this terminates NVTX range.

    Check torch.cuda.nvtx.range_pop's document for more details.
    """
    if torch.cuda.is_available():
        torch.cuda.nvtx.range_pop()
```